nowcasting_dataset/manager.py

```diff
@@ -0,0 +1,384 @@
+                freq='5 minutes' then every datetime will be at 00, 05, ..., 55 minutes
+                past the hour.
+
+        Returns:  Valid t0 datetimes, taking into consideration all DataSources,
+            filtered by daylight hours (SatelliteDataSource.datetime_index() removes the night
+            datetimes).
+        """
+        logger.debug(
+            f"Getting the intersection of time periods across all DataSources at freq={freq}..."
+        )
+        if set(self.data_sources.keys()) != set(ALL_DATA_SOURCE_NAMES):
+            logger.warning(
+                "Computing available t0 datetimes using less than all available DataSources!"
+                " Are you sure you mean to do this?!"
+            )
+
+        # Get the intersection of t0 time periods from all data sources.
+        t0_time_periods_for_all_data_sources = []
+        for data_source_name, data_source in self.data_sources.items():
+            logger.debug(f"Getting t0 time periods for {data_source_name}")
+            try:
+                t0_time_periods = data_source.get_contiguous_t0_time_periods()
+            except NotImplementedError:
+                # Skip data_sources with no concept of time.
+                logger.debug(f"Skipping {data_source_name} because it has not concept of datetime.")
+            else:
+                t0_time_periods_for_all_data_sources.append(t0_time_periods)
+
+        intersection_of_t0_time_periods = nd_time.intersection_of_multiple_dataframes_of_periods(
+            t0_time_periods_for_all_data_sources
+        )
+
+        t0_datetimes = nd_time.time_periods_to_datetime_index(
+            time_periods=intersection_of_t0_time_periods, freq=freq
+        )
+        logger.debug(
+            f"Found {len(t0_datetimes):,d} datetimes at freq=`{freq}` across"
+            f" DataSources={list(self.data_sources.keys())}."
+            f"  From {t0_datetimes[0]} to {t0_datetimes[-1]}."
+        )
+        return t0_datetimes
+
+    def sample_spatial_and_temporal_locations_for_examples(
+        self, t0_datetimes: pd.DatetimeIndex, n_examples: int
+    ) -> pd.DataFrame:
+        """
+        Computes the geospatial and temporal locations for each training example.
+
+        The first data_source in this DataSourceList defines the geospatial locations of
+        each example.
+
+        Args:
+            t0_datetimes: All available t0 datetimes.  Can be computed with
+                `DataSourceList.get_t0_datetimes_across_all_data_sources()`
+            n_examples: The number of examples requested.
+
+        Returns:
+            Each row of each the DataFrame specifies the position of each example, using
+            columns: 't0_datetime_UTC', 'x_center_OSGB', 'y_center_OSGB'.
+        """
+        shuffled_t0_datetimes = np.random.choice(t0_datetimes, size=n_examples)
+        # TODO: Speed this up by splitting the shuffled_t0_datetimes across multiple processors.
```

Comment:
is that needed? how much work / time does this take?

Reply:
Good question.  This is definitely low priority.  I've added some more context to the comment.

`sample_spatial_and_temporal_locations_for_examples` currently takes about half an hour for 25,000 batches.  Which is a bit long.  But I think the better fix is probably #305.